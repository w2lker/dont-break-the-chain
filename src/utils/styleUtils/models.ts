export type cssAttribute = 'unset' | 'null' | number | string | undefined | null;

type TArrayWithLength<TItem, TLength extends number> = [TItem, ...TItem[]] & { length: TLength };
export type BoundAttributeArray = TArrayWithLength<string, 1| 2 | 3 | 4>;
