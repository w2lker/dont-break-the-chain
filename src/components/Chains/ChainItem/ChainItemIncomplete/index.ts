import ChainItemIncomplete from './ChainItemIncomplete';
export default ChainItemIncomplete;
