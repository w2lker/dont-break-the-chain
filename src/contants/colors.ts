import colorBlue from '@material-ui/core/colors/blue';
import colorBrown from '@material-ui/core/colors/brown';
import colorCyan from '@material-ui/core/colors/cyan';
import colorDeepPurple from '@material-ui/core/colors/deepPurple';
import colorGreen from '@material-ui/core/colors/green';
import colorIndigo from '@material-ui/core/colors/indigo';
import colorOrange from '@material-ui/core/colors/orange';
import colorPink from '@material-ui/core/colors/pink';
import colorPurple from '@material-ui/core/colors/purple';
import colorTeal from '@material-ui/core/colors/teal';

export type colorsKeys = 'pink' | 'purple' | 'deepPurple' | 'indigo' | 'blue' | 'cyan' | 'teal' | 'green' | 'orange' | 'brown';

export const colors: {[color: string]: any}  = {
  blue: colorBlue,
  brown: colorBrown,
  pink: colorPink,
  cyan: colorCyan,
  deepPurple: colorDeepPurple,
  indigo: colorIndigo,
  green: colorGreen,
  orange: colorOrange,
  purple: colorPurple,
  teal: colorTeal,
};
