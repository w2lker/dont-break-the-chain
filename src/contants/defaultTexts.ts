const defaultTexts = {
  navigationMenu: {
    chains: 'Chains',
    habits: 'Habits',
    profile: 'Profile',
  },
  habits: {
    header: {
      title: 'My Habits',
    },
  },
  profile: {
    header: {
      title: 'Profile',
    },
    stats: {
      total: {
        description: 'points received',
      },
      days: {
        subtitle: 'days',
        description: 'longest chain',
      },
    },
  },
};

export default defaultTexts;
