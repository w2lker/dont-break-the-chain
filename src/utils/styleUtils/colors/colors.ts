import {bonWarning} from '../helpers/warning/warning';

export interface IRgbColor {
  red: number;
  green: number;
  blue: number;
}

const black: IRgbColor = {
  red: 0,
  green: 0,
  blue: 0,
};

function regularHexToRgb(hex: string): IRgbColor {
  const [red, green, blue] = [0, 2, 4]
    .map(offset => hex.substring(offset, offset + 2))
    .map(hexCh => parseInt(hexCh, 16));
  return { red, green, blue };
}

export function hexToRgbMapped(hex?: string): IRgbColor {
  const warnType = 'HEX to RGB converter';
  if (!hex) {
    return black;
  }
  if (hex.length === 1) {
    return regularHexToRgb(hex.repeat(6));
  }

  if (hex.length === 2) {
    return regularHexToRgb(hex.repeat(3));
  }

  if (hex.length === 3) {
    const [r, g, b] = hex.split('') as string[];
    return regularHexToRgb(`${r}${r}${g}${g}${b}${b}`);
  }

  if (hex.length === 4 || hex.length === 5) {
    const hex3 = hex.slice(0, 3);
    // tslint:disable-next-line:no-console
    bonWarning(`can't verify converting from HEX to RGB — received HEX value with ${hex.length} digits. It is parsed as ${hex3} value`, warnType);
    return hexToRgbMapped(hex3);
  }

  if (hex.length > 6) {
    const hex6 = hex.slice(0, 6);
    // tslint:disable-next-line:no-console
    bonWarning(`can't verify converting from HEX to RGB — received HEX value with ${hex.length} digits. It is parsed as ${hex6} value`, warnType);
    return regularHexToRgb(hex6);
  }

  return regularHexToRgb(hex);
}
