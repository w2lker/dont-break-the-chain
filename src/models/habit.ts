import {List} from 'immutable';
import {colorsKeys} from "../contants/colors";

export const habitStatus = {
  incomplete: 0,
  complete: 1,
  paused: 2,
  start: 3,
  end: 4,
  incompleteToday: 5
};

export interface IHabitDate {
  date?: any;
  status?: string;
  score?: number;
}

export interface IHabit {
  id?: number;
  name?: string;
  color?: colorsKeys;
  icon?: string;
  currentChain?: number;
  longestChain?: number;
  dates?: List<IHabitDate>;
}