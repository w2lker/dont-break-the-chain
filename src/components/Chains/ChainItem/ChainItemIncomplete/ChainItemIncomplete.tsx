import React from "react";

import MaterialDesignIcon from "../../../basic/Material-Icon";
import IconButton from "@material-ui/core/IconButton";

import {colors, colorsKeys} from "../../../../contants/colors";
import {habitStatus, habitStatusesType} from "../../../../models/habit";


export interface IChainItemIncompleteProps {
  status: habitStatusesType;
  color: colorsKeys;
  onAdd: () => void;
}

const ChainItemIncomplete: React.FC<IChainItemIncompleteProps> = (props) => {
  const {status, onAdd} = props;

  if (status !== habitStatus.incompleteToday) {
    return null;
  }

  const color = colors[props.color || 'blue'][500];

  return (
    <IconButton
        onClick={onAdd}
    >
        <MaterialDesignIcon
            name="plus-circle"
            style={{ color }}
        />
    </IconButton>
  );
};

export default ChainItemIncomplete;