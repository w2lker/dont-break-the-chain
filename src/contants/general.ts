import { RequestStatuses } from '../models/general';

export const requestStatuses = {
  active: 'active',
  success: 'success',
  error: 'error',
};
