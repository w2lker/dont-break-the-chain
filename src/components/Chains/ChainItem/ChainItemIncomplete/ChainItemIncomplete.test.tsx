import React from "react";
import { shallow, mount, render} from "enzyme";
import ChainItemIncomplete, {IChainItemIncompleteProps} from "./ChainItemIncomplete";
import {habitStatus} from "../../../../models/habit";
import {colors, colorsKeys} from "../../../../contants/colors";

describe('Chain Item Incomplete', () => {
  const emptyProps: IChainItemIncompleteProps = {
    // @ts-ignore
    status: undefined,
    // @ts-ignore
    color: undefined,
    onAdd: () => {},
  };

  const sampleProps: IChainItemIncompleteProps = {
    status: habitStatus.incompleteToday,
    color: "deepPurple",
    onAdd: () => {},
  };

  it('match snapshot', () => {
      const component = mount(<ChainItemIncomplete {...sampleProps} />);
      expect(component.debug()).toMatchSnapshot();
  });

  it('skips render with empty props passed', () => {
      const component = shallow(<ChainItemIncomplete {...emptyProps} />);
      expect(component.isEmptyRender()).toBe(true);
  });

  it('skip render with status !== incompleteToday', () => {
    const otherStatuses = Object.values(habitStatus).filter( item => item !== habitStatus.incompleteToday );
    otherStatuses.forEach( (status) => {
      const newProps: IChainItemIncompleteProps = {
        ...sampleProps,
        // @ts-ignore
        status: status,
      };
      const component = shallow(<ChainItemIncomplete {...newProps} />);
      expect(component.isEmptyRender()).toBe(true);
    })
  });

  describe('testing color assignment', () => {
    function testColorAssignment(color: colorsKeys) {
      const newProps: IChainItemIncompleteProps = {
        ...sampleProps,
        // @ts-ignore
        color: color,
      };

      const sampleStyle = {
        color: colors[color || 'blue'][500],
      };

      const component = shallow(<ChainItemIncomplete {...newProps} />);
      const iconComponent = component.find('MaterialDesignIcon');
      expect(iconComponent.props().style).toEqual(sampleStyle);
    }

    it('renders with default color', () => testColorAssignment('blue'));
    it('renders with provided color', () => {
      const colorKeys = Object.keys(colors);
      // @ts-ignore
      colorKeys.forEach(testColorAssignment);
    })
  });

  it('fires click callback', () => {
    const mockAddClick = jest.fn();
    const newProps: IChainItemIncompleteProps = {
      ...sampleProps,
      onAdd: mockAddClick,
    };
    const component = shallow(<ChainItemIncomplete {...newProps} />);
    component.simulate('click');
    expect(mockAddClick.mock.calls.length).toEqual(1);
  })
});