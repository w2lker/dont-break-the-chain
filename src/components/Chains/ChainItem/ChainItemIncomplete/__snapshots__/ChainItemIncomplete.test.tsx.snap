// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Chain Item Incomplete match snapshot 1`] = `
"<ChainItemIncomplete status={5} color=\\"deepPurple\\" onAdd={[Function: onAdd]}>
  <WithStyles(ForwardRef(IconButton)) onClick={[Function: onAdd]}>
    <ForwardRef(IconButton) classes={{...}} onClick={[Function: onAdd]}>
      <WithStyles(ForwardRef(ButtonBase)) className=\\"MuiIconButton-root\\" centerRipple={true} focusRipple={true} disabled={false} onClick={[Function: onAdd]}>
        <ForwardRef(ButtonBase) classes={{...}} className=\\"MuiIconButton-root\\" centerRipple={true} focusRipple={true} disabled={false} onClick={[Function: onAdd]}>
          <button className=\\"MuiButtonBase-root MuiIconButton-root\\" onBlur={[Function]} onClick={[Function: onAdd]} onFocus={[Function]} onKeyDown={[Function]} onKeyUp={[Function]} onMouseDown={[Function]} onMouseLeave={[Function]} onMouseUp={[Function]} onDragLeave={[Function]} onTouchEnd={[Function]} onTouchMove={[Function]} onTouchStart={[Function]} tabIndex={0} type=\\"button\\" disabled={false}>
            <span className=\\"MuiIconButton-label\\">
              <MaterialDesignIcon name=\\"plus-circle\\" style={{...}}>
                <span className=\\"material-icon mdi mdi-plus-circle\\" style={{...}} />
              </MaterialDesignIcon>
            </span>
            <NoSsr>
              <WithStyles(undefined) center={true}>
                <ForwardRef(TouchRipple) classes={{...}} center={true}>
                  <span className=\\"MuiTouchRipple-root\\">
                    <TransitionGroup component={{...}} exit={true} childFactory={[Function: childFactory]} />
                  </span>
                </ForwardRef(TouchRipple)>
              </WithStyles(undefined)>
            </NoSsr>
          </button>
        </ForwardRef(ButtonBase)>
      </WithStyles(ForwardRef(ButtonBase))>
    </ForwardRef(IconButton)>
  </WithStyles(ForwardRef(IconButton))>
</ChainItemIncomplete>"
`;
